test uv_barrier_t size in right block

semarie opened the pull request:
the local uv_barrier_t implementation is used: for AIX, OpenBSD and any system not defining PTHREAD_BARRIER_SERIAL_THREAD.

as the uv_barrier_t struct was modified and it isn't a pointer anymore, the check for uv_barrier_t size isn't right anymore for systems not using local uv_barrier_t implementation. so move the check inside the block implementing uv_barrier_t using pthread_barrier_t.

tested on OpenBSD, where the build failed due to the STATIC_ASSERT().

Cherry-pick #35 for the master branch.

tkelman commented:
Should `no_epoll_wait` be initialized? Otherwise lgtm

yuyichao commented:
I think this is not necessary given it is a static variable. (I couldn't find it in C89 standard but it required by c++ and c99). It's better practice in general to initialize them but we should probably just stick with what libuv upstream has.

yuyichao commented:
Will merge sometime later today (12/30).

Should I just open an `backport pending 0.4` PR against base julia that update the commit SHA as well as the tarball checksums?

tkelman commented:
We're on a slightly different libuv branch for 0.4, there were a few commits I didn't feel were well tested enough for the release branch. Does release 0.4 even build on aarch64?

yuyichao commented:
I'm assuming that given JuliaLang/julia#14373 was reported for 0.4.2, it should have gone through the first stage bootstrap just fine. I'll check if this patch is enough to complete the compilation.

yuyichao commented:
Yes I can get a REPL on release-0.4 with this patch applied on libuv.

```
julia> versioninfo()
Julia Version 0.4.3-pre+6
Commit adffe19 (2015-12-11 00:38 UTC)                                             
Platform Info:                                                                    
  System: Linux (aarch64-unknown-linux-gnu)                                       
  CPU: unknown                                                                    
  WORD_SIZE: 64                                                                   
  BLAS: libblas                                                                   
  LAPACK: liblapack                                                               
  LIBM: libm                                                                      
  LLVM: libLLVM-3.7.0                                                             
```

semarie committed:
test uv_barrier_t size in right block

the local uv_barrier_t implementation is used: for AIX, OpenBSD and any system
not defining PTHREAD_BARRIER_SERIAL_THREAD.

as the uv_barrier_t struct was modified and it isn't a pointer anymore, the
check for uv_barrier_t size isn't right anymore for systems not using local
uv_barrier_t implementation. so move the check inside the block implementing
uv_barrier_t using pthread_barrier_t.

tested on OpenBSD, where the build failed due to the STATIC_ASSERT().

semarie commented:
@vtjnash now I have a doubt. you were speaking about the master branch here ? or the master branch in upstream ?
